plugins/receivers/lark/client.go

```diff
@@ -0,0 +1,223 @@
+package lark
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/goto/siren/pkg/errors"
+	"github.com/goto/siren/pkg/httpclient"
+	"github.com/goto/siren/pkg/retry"
+	lark "github.com/larksuite/oapi-sdk-go/v3"
+	larkcore "github.com/larksuite/oapi-sdk-go/v3/core"
+	larkcontact "github.com/larksuite/oapi-sdk-go/v3/service/contact/v3"
+	larkim "github.com/larksuite/oapi-sdk-go/v3/service/im/v1"
+)
+
+const (
+	defaultLarkAPIHost = "https://open.larksuite.com"
+	larkPathOAuth      = "/open-apis/auth/v3/tenant_access_token/internal/"
+)
+
+type Channel struct {
+	ID   string `json:"id"`
+	Name string `json:"name"`
+}
+
+type ClientOption func(*Client)
+
+// ClientWithHTTPClient assigns custom http client when creating a lark client
+func ClientWithHTTPClient(httpClient *httpclient.Client) ClientOption {
+	return func(c *Client) {
+		c.httpClient = httpClient
+	}
+}
+
+// ClientWithRetrier wraps client call with retrier
+func ClientWithRetrier(runner retry.Runner) ClientOption {
+	return func(c *Client) {
+		// note: for now retry only happen in send message context method
+		c.retrier = runner
+	}
+}
+
+type Client struct {
+	cfg        AppConfig
+	httpClient *httpclient.Client
+	retrier    retry.Runner
+}
+
+// NewClient is a constructor to create lark client.
+// this version uses lark v3 SDK.
+func NewClient(cfg AppConfig, opts ...ClientOption) *Client {
+	c := &Client{
+		cfg: cfg,
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	if cfg.APIHost == "" {
+		c.cfg.APIHost = defaultLarkAPIHost
+	}
+
+	// sanitize
+	c.cfg.APIHost = c.cfg.APIHost + "/"
+
+	if c.httpClient == nil {
+		c.httpClient = httpclient.New(cfg.HTTPClient)
+	}
+
+	return c
+}
+
+// GetWorkspaceChannels fetches list of joined channel of a client
+func (c *Client) GetWorkspaceChannels(ctx context.Context, clientID, clientSecret string) ([]Channel, error) {
+	var client = lark.NewClient(clientID, clientSecret)
```

Comment:
Is the `clientID` and `clientSecret` is specific for a certain bot in lark? 

I don't think we need to always create a client every time this function is being called. We should only create a client once during `NewClient` constructor.

Reply:
yes clientID and clientSecret are specific for a lark bot.

Reply:
Yes we can do it in constructor. And we can use env vars of clientID and clientSecret, then no need to store in receiver config if we would support only a single bot.

Reply:
Oh okay, if it is a specific bot then I guess it is fine doing it so.